[feat] 페인팅 기능 구현

HyewonKkang opened the pull request:
## 상세내용
- 드로잉 툴의 페인팅 기능을 구현했어요. 기존에 공유했던 일부 색이 안채워지는 이슈나 클릭 위치 외부 공간이 채워지는 이슈를 해결했어요.
- ~~현재는 isFilling과 isPainting state를 하나의 state로 수정할 계획이에요.~~ 수정해서 커밋했어요.

HyewonKkang committed:
feat: [FE] 채우기 기본 기능 구현

- Intent

HyewonKkang committed:
refactor: [FE] setPixel() 위치 수정

- Intent
useCanvas() 내 state에 의존하지 않은 함수 utils로 옮김

HyewonKkang committed:
fix: [FE] 채우기 오류 해결

- Intent
클릭위치 외의 공간이 채워지는 이슈 및 특정 색이 안채워지는 이슈 해결

HyewonKkang committed:
fix: [FE] rest 네이밍 수정

rest -> restProps로 변경
- Intent
rest 네이밍 의미 파악이 어려움

yscriuf commented on frontend/src/hooks/useCanvas.ts:
```diff
@@ -19,11 +19,11 @@ interface Coordinate {
     const [pos, setPos] = useState<Coordinate | undefined>({ x: 0, y: 0 });
     const [isPainting, setIsPainting] = useState<boolean>(false);
     const [isFilling, setIsFilling] = useState<boolean>(false);
-    const [color, setColor] = useState<Uint8ClampedArray>(convertHexToRgba('#001D2E'));
```
State에서 Ref로 수정하신 이유가 궁금해요!

HyewonKkang replied:
우선 color 값이 달라진다고해서 컴포넌트가 리렌더링될 필요가 없었고, 단순히 현재 색상 값을 저장할 용도로만 사용해도 될 것으로 생각해 useRef를 이용했어요. 그리고 state가 즉각적으로 업데이트 되지 않아 생기는 이슈 때문에 수정했어요:)

jayjeong8 commented on frontend/src/hooks/usePalette.ts:
```diff
@@ -61,7 +61,7 @@ function usePalette({
     const onClickColor = (color: string) => {
         setSelectedColor(color);
-        setSelectedTool(ToolsType.PEN);
+        if (selectedTool === ToolsType.ERASER) setSelectedTool(ToolsType.PEN);
```
지우개를 쓰다가 색을 선택하면 다시 펜이 되는거 좋아요!! 다른 게임에서 불편하다고 느낀 기억이 있는데 자동으로 펜이 선택된다니 좋네요 :)

HyewonKkang replied:
맞아요! 디버깅 하면서 확인할때도 불편해서 넣었어요.. 감사해요 ㅎㅎ

jayjeong8 commented on frontend/src/hooks/useCanvas.ts:
```diff
@@ -19,11 +19,11 @@ interface Coordinate {
 function useCanvas() {
     const canvasRef = useRef<HTMLCanvasElement>(null);
     const ctxRef = useRef<any>(null);
+    const curColor = useRef<Uint8ClampedArray>(convertHexToRgba('#001D2E'));
```
페인트통을 부었을 때 옆 픽셀과의 값 비교를 위해 hex를 rgba로 변환하는 것인지 궁금해요! 그리고 Uint8ClampedArray를 사용하는 이유도 궁금합니다!

yscriuf replied:
> 페인트통을 부었을 때 옆 픽셀과의 값 비교를 위해 hex를 rgba로 변환하는 것인지 궁금해요! 그리고 Uint8ClampedArray를 사용하는 이유도 궁금합니다!

Uint8ClampedArray를 사용하는 이유는 [아티클](https://luxuriant-sound-22a.notion.site/e08f88b27df04591800ffea5fa7747e0)에 나와있더라구요!

jayjeong8 replied:
오 Uint8ClampedArray를 쓰면 0~255의 값만 넣을 수 있고 그 밖의 값은 자동으로 0이나 255로 바뀌는군요!!

HyewonKkang replied:
해당 타입을 이용한 이유는 캔버스의 픽셀 데이터가 ImageData 객체안의 data 속성으로 저장되어 있는데 그때 타입이랑 그대로 일치시켜서 이용하려고 Uint8ClampedArray를 이용했어요.
추가로 말씀하신 0~255 범위 맞아요! 처음에 일부 색이 안나온다고 했던 이유도 실제 rgba의 알파 값은 0-1 사이 값인데 Uint8ClampedArray의 알파 값은 0-255 범위여서 생긴 이슈였어요 ㅎㅎ

yscriuf commented on frontend/src/utils/canvas.ts:
```diff
@@ -0,0 +1,31 @@
+    return new Uint8ClampedArray(rgba);
+};
+
+export const isValidSquare = (imageData: any, x: number, y: number) => {
+    return x >= 0 && x < imageData.width && y >= 0 && y < imageData.height;
+};
+
+export const getPixelOffset = (imageData: any, x: number, y: number) => {
+    return (y * imageData.width + x) * 4;
+};
+
+export const getPixelColor = (imageData: any, x: number, y: number) => {
+    if (isValidSquare(imageData, x, y)) {
+        const offset = getPixelOffset(imageData, x, y);
+        return imageData.data.slice(offset, offset + 4);
+    } else {
+        return [-1, -1, -1, -1]; // invalid color
+    }
+};
```
엄청난 고민을 한 흔적이 보이는데... 고생하셨어요!!!! 👍

jayjeong8 commented on frontend/src/hooks/useCanvas.ts:
```diff
@@ -19,11 +19,11 @@ interface Coordinate {
     const [pos, setPos] = useState<Coordinate | undefined>({ x: 0, y: 0 });
     const [isPainting, setIsPainting] = useState<boolean>(false);
     const [isFilling, setIsFilling] = useState<boolean>(false);
-    const [color, setColor] = useState<Uint8ClampedArray>(convertHexToRgba('#001D2E'));
```
useState에서 useRef로 바뀌었는데 이게 오류의 문제였다면 어떤게 문제였는지 알고싶어요!

HyewonKkang replied:
https://github.com/boostcampwm-2022/web37-ZelloDraw/pull/31#discussion_r1033107055 
지현님께서 비슷한 질문 남겨주셔서 코멘트 남겼어요!
오류는 작성한대로 state가 바로 update되지 않아 색을 바꿔도 바로 반영되지 않는 문제가 있었어요!

HyewonKkang committed:
fix: [FE] rgba 알파값 매핑 함수 수정

jayjeong8 commented:
잘 모르는 기능도 보이고 어떻게 구현했을지 궁금해져서 기술공유 시간에 혜원님이 해줄 얘기가 기대되어요!! 고생했어요~~!

HyewonKkang committed:
fix: [FE] drawState로 현재 그리기 도구 상태 저장

- Intent
기존의 isFilling, isPainting 두 state를 drawState로 통일시킴
페인트 도구를 선택한 상태로 선이 그려지던 버그 수정

LouiMinister commented on frontend/src/hooks/useCanvas.ts:
```diff
@@ -36,22 +40,67 @@ function useCanvas() {
 
+    const floodFill = (x: number, y: number, fillColor: Uint8ClampedArray) => {
+        const imageData = ctxRef.current.getImageData(0, 0, CANVAS_WIDTH, CANVAS_HEIGHT);
+        const visited = new Uint8Array(imageData.width, imageData.height);
```
일반 배열이 아니라 Unit8Array를 쓴 이유가 있나요?

HyewonKkang committed:
Merge branch 'develop' into feat/painting
